plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/jvm/SerializableCodegenImpl.kt

```diff
@@ -261,12 +277,94 @@ class SerializableCodegenImpl(
             return 0 to propStartVar
         } else {
             val superProps = bindingContext.serializablePropertiesFor(superClass).serializableProperties
-            val creator = buildInternalConstructorDesc(propStartVar, 1, classCodegen, superProps)
+            val creator = buildInternalConstructorDesc(descVar, propStartVar, maskVar, classCodegen, superProps)
             invokespecial(superType, "<init>", creator, false)
             return superProps.size to propStartVar + superProps.sumBy { it.asmType.size }
         }
     }
 
+    private fun InstructionAdapter.generateNewGoldenMaskCheck(descVar: Int, maskVar: Int) {
+        val allPresentsLabel = Label()
+        val maskSlotCount = properties.serializableProperties.bitMaskSlotCount()
+        if (maskSlotCount == 1) {
+            var goldenMask = 0
+            var requiredBit = 1
+            for (property in properties.serializableProperties) {
+                if (!property.optional) {
+                    goldenMask = goldenMask or requiredBit
+                }
+                requiredBit = requiredBit shl 1
+            }
+
+            iconst(goldenMask)
+            dup()
+            load(maskVar, OPT_MASK_TYPE)
+            and(OPT_MASK_TYPE)
+            ificmpeq(allPresentsLabel)
```

Comment:
It looks like this solution does not work if we have optional fields. In this case, seen mask contains more one bits than the golden one, and therefore they're not equal.

Reply:
Seen mask can't contains more bits because first run "and" operation and only then comparing with golden mask: (seen and golden) == golden

Reply:
hm, indeed!